### ISSUE TITLE
Excluded paths in nested configurations are ignored

### ISSUE THREAD

**Issue by fixpunkt** [realm/SwiftLint#2447]
### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

My example project has this structure:

```
.
├── .swiftlint.yml
└── A
    ├── .swiftlint.yml
    ├── B
    │   └── Longcat.swift
    └── Hello.swift
```

`./swiftlint.yml` is empty, `./A/swiftlint.yml` has the following contents:

```yaml
excluded:
  - B
```

When I run `swiftlint` from the project's root directory, `Longcat.swift` gets linted although it should be excluded.

An explanation as to why I'm trying to do this: We have a common base `.swiftlint.yml` for our apps, which we symlink from a CocoaPod. Since we need to add project-specific excludes as well (and SwiftLint unfortunately doesn't seem to support shared base configurations at all), I need to mark individual folders as excluded from nested configurations.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths
Linting 'Longcat.swift' (1/2)
Linting 'Hello.swift' (2/2)
.../swiftlint-example/A/B/Longcat.swift:1: warning: Line Length Violation: Line should be 120 characters or less: currently 189 characters (line_length)
Done linting! Found 1 violation, 0 serious in 2 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?

`0.27.0`

* Installation method used (Homebrew, CocoaPods, building from source, etc)?

In this instance, I was using the `swiftlint` from `brew install swiftlint`

* Paste your configuration file:

See above.

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

Yes. Toplevel is empty, nested config is the one shown above.

* Which Xcode version are you using (check `xcode-select -p`)?

`Version 10.0 (10A255)`

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

I've attached my minimal example project: [swiftlint-nested-exclude-example.zip](https://github.com/realm/SwiftLint/files/2526054/swiftlint-nested-exclude-example.zip)

**Comment by ZevEisenberg** [comment]
+1 I believe I'm seeing a similar issue.

**Comment by Bruschidy54** [comment]
I think I found a solution. When SwiftLint is linting files in visitLintableFiles in Configuration+CommandLine, it:

1. Gathers all lintable files in getFiles. This is where the exclusion of files occurs based on the parent configuration's exclusion list.

2. These files are grouped by their specific configuration in groupFiles. This is where configurations for each available file are determined (and if nested configurations exist, merged). After these configurations are determined and the files are grouped accordingly, no more files are excluded from the lintable files list, generating the bug where excluded lists of files and directories are ignored in nested configurations.

3. Finally, each file is visited by the linter.

My solution is to skip files whose merged configurations specify they should be excluded in step 2 or groupFiles. Therefore, they will not be visited in step 3.

Take a look at my PR :)
https://github.com/realm/SwiftLint/pull/2621
